src/index/DeltaTriples.cpp

```diff
@@ -0,0 +1,189 @@
+        // integration is done
+        idTriples, perm.metaData().blockData(), perm.keyOrder(), shouldExist,
+        cancellationHandle);
+    cancellationHandle->throwIfCancelled();
+    intermediateHandles[permutation] =
+        getLocatedTriplesPerBlock(permutation).add(locatedTriples);
+    cancellationHandle->throwIfCancelled();
+  }
+  std::vector<DeltaTriples::LocatedTripleHandles> handles{idTriples.size()};
+  for (auto permutation : Permutation::ALL) {
+    for (size_t i = 0; i < idTriples.size(); i++) {
+      handles[i].forPermutation(permutation) =
+          intermediateHandles[permutation][i];
+    }
+  }
+  return handles;
+}
+
+// ____________________________________________________________________________
+void DeltaTriples::eraseTripleInAllPermutations(
+    DeltaTriples::LocatedTripleHandles& handles) {
+  // Erase for all permutations.
+  for (auto permutation : Permutation::ALL) {
+    auto ltIter = handles.forPermutation(permutation);
+    getLocatedTriplesPerBlock(permutation).erase(ltIter->blockIndex_, ltIter);
+  }
+}
+
+// ____________________________________________________________________________
+void DeltaTriples::insertTriples(
+    ad_utility::SharedCancellationHandle cancellationHandle,
+    std::vector<IdTriple<0>> triples) {
+  LOG(DEBUG) << "Inserting " << triples.size()
+             << " triples (including idempotent triples)." << std::endl;
+  std::ranges::sort(triples);
+  // Unique moves all duplicate items to the end and returns iterators for that
+  // subrange.
+  auto [first, last] = std::ranges::unique(triples);
+  triples.erase(first, last);
+  std::erase_if(triples, [this](const IdTriple<0>& triple) {
+    return triplesInserted_.contains(triple);
+  });
+  std::ranges::for_each(triples, [this](const IdTriple<0>& triple) {
+    auto handle = triplesDeleted_.find(triple);
+    if (handle != triplesDeleted_.end()) {
+      eraseTripleInAllPermutations(handle->second);
+      triplesDeleted_.erase(triple);
+    }
+  });
+
+  std::vector<LocatedTripleHandles> handles =
+      locateAndAddTriples(std::move(cancellationHandle), triples, true);
+
+  AD_CORRECTNESS_CHECK(triples.size() == handles.size());
+  // TODO<qup42>: replace with std::views::zip in C++23
+  for (size_t i = 0; i < triples.size(); i++) {
+    triplesInserted_.insert({triples[i], handles[i]});
+  }
+}
+
+// ____________________________________________________________________________
+void DeltaTriples::deleteTriples(
+    ad_utility::SharedCancellationHandle cancellationHandle,
+    std::vector<IdTriple<0>> triples) {
+  LOG(DEBUG) << "Deleting " << triples.size()
+             << " triples (including idempotent triples)." << std::endl;
+  std::ranges::sort(triples);
+  auto [first, last] = std::ranges::unique(triples);
+  triples.erase(first, last);
+  std::erase_if(triples, [this](const IdTriple<0>& triple) {
+    return triplesDeleted_.contains(triple);
```

Comment:
This is 90% code duplication with the previous function.
Factor out the common part (basically the "inserted"and "deleted" triples are switched, "true " becomes "false", and the log has a different beginning word, the rest is the same).

This will be very useful to locally change the implementation of those data structures ( I am not yet fully convinced of the `Ordere_Map<T>::iterator` approach:)